src/combinatorics/simple_graph.lean

```diff
@@ -0,0 +1,202 @@
+/-
+Copyright (c) 2020 Aaron Anderson, Jalex Stark, Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Aaron Anderson, Jalex Stark, Kyle Miller.
+-/
+import data.fintype.basic
+import data.sym2
+/-!
+# Simple graphs
+
+This module defines simple graphs on a vertex type `V` as an
+irreflexive symmetric relation.
+
+There is a basic API for locally finite graphs and for graphs with
+finitely many vertices.
+
+## Main definitions
+
+* `simple_graph` is a structure for symmetric, irreflexive relations
+
+* `neighbor_set` is the `set` of vertices adjacent to a given vertex
+
+* `neighbor_finset` is the `finset` of vertices adjacent to a given vertex,
+   if `neighbor_set` is finite
+
+## Implementation notes
+
+* A locally finite graph is one with instances `∀ v, fintype (G.neighbor_set v)`.
+
+* Given instances `decidable_rel G.adj` and `fintype V`, then the graph
+is locally finite, too.
+
+TODO: This is the simplest notion of an unoriented graph.  This should
+eventually fit into a more complete combinatorics hierarchy which
+includes multigraphs and directed graphs.  We begin with simple graphs
+in order to start learning what the combinatorics hierarchy should
+look like.
+
+TODO: Part of this would include defining, for example, subgraphs of a
+simple graph.
+
+-/
+open finset
+universe u
+
+/--
+A simple graph is an irreflexive symmetric relation `adj` on a vertex type `V`.
+The relation describes which pairs of vertices are adjacent.
+There is exactly one edge for every pair of adjacent edges;
+see `simple_graph.edge_set` for the corresponding edge set.
+-/
+structure simple_graph (V : Type u) :=
+(adj : V → V → Prop)
+(sym : symmetric adj . obviously)
+(loopless : irreflexive adj . obviously)
+
+/--
+The complete graph on a type `V` is the simple graph with all pairs of distinct vertices adjacent.
+-/
+def complete_graph (V : Type u) : simple_graph V :=
+{ adj := ne }
+
+instance (V : Type u) : inhabited (simple_graph V) :=
+⟨complete_graph V⟩
+
+instance complete_graph_adj_decidable (V : Type u) [decidable_eq V] :
+  decidable_rel (complete_graph V).adj :=
+by { dsimp [complete_graph], apply_instance }
+
+namespace simple_graph
+variables {V : Type u} (G : simple_graph V)
+
+/-- `G.neighbor_set v` is the set of vertices adjacent to `v` in `G`. -/
+def neighbor_set (v : V) : set V := set_of (G.adj v)
+
+lemma ne_of_adj {a b : V} (hab : G.adj a b) : a ≠ b :=
+by { rintro rfl, exact G.loopless a hab }
+
+/--
+The edges of G consist of the unordered pairs of vertices related by
+`G.adj`.
+-/
+def edge_set : set (sym2 V) := sym2.from_rel G.sym
+
+@[simp]
+lemma edge_iff_adj {v w : V} : ⟦(v, w)⟧ ∈ G.edge_set ↔ G.adj v w :=
+by refl
+
+lemma adj_iff_exists_edge {v w : V} :
+  G.adj v w ↔ v ≠ w ∧ ∃ (e ∈ G.edge_set), v ∈ e ∧ w ∈ e :=
```

Comment:
The equivalence `G.adj v w ↔ ∃ (e ∈ G.edge_set), v ∈ e ∧ w ∈ e` also holds, right? That might be a simpler lemma (and together with `ne_of_adj` you still get both components of the RHS).

Reply:
I have to periodically remind myself about why this needs the `v ≠ w`: it's because if `v = w`, then `∃ (e ∈ G.edge_set), v ∈ e ∧ w ∈ e` is satisfied by every edge incident to `v`.

Reply:
Ah, you are right! Given that this has confused me (and apparently also you multiple times), do you mind adding that as a docstring to the lemma?

Reply:
Done